lib/abide_dev_utils/cem/generate/reference.rb

```diff
@@ -262,6 +262,40 @@ def generate!
             resource_reference_builder
           end
 
+          # This function act as a filter for controls based on the profile and level selections.
+          # There are few scanarios that can happen:
+          # 1. If no selections are made for profile or level, then all profiles and levels of control will be selected.
+          # 2. If selections are made for profile, then only the selected profile and all levels of control will be selected.
+          # 3. If selections are made for level, then only the selected level and all profiles of control will be selected.
+          # This function adds in some runtime overhead because we're checking each control's level and profile which is
+          # what we're going to be doing later when building the level and profile markdown, but this is
+          # necessary to ensure that the reference.md is generated the way we want it to be.
+          def verify_profile_and_level_selections
+            return true if @opts[:select_profile].nil? && @opts[:select_level].nil?
+
+            if @opts[:select_profile].nil? && !@opts[:select_level].nil?
+              @control.levels.each do |level|
+                return true if control_level_filter(level)
+              end
+            elsif !@opts[:select_profile].nil? && @opts[:select_level].nil?
+              @control.profiles.each do |profile|
+                return true if control_profile_filter(profile)
+              end
+            elsif !@opts[:select_profile].nil? && !@opts[:select_level].nil?
+              contain_level = false
+              contain_profile = false
+
+              @control.levels.each do |level|
+                contain_level = true if control_level_filter(level)
+              end
+              @control.profiles.each do |profile|
+                contain_profile = true if control_profile_filter(profile)
+              end
+
+              return true if contain_level && contain_profile
+            end
+          end
+
```

Comment:
Since we're checking this here, could we return the valid level and profile from this function and pass them into `#generate!`? Then we could skip checking the level and profile during generation

Reply:
Updated so that we skip checking during the generation process.